pkg/ts/server.go

```diff
@@ -78,28 +105,105 @@ func (s *Server) Query(
+	// in the same order as the queries in the request.
+	type queryResultWithError struct {
+		queryNum int
+		result   tspb.TimeSeriesQueryResponse_Result
+		err      error
 	}
-	for _, query := range request.Queries {
-		datapoints, sources, err := s.db.Query(
-			ctx,
-			query,
-			Resolution10s,
-			sampleNanos,
-			request.StartNanos,
-			request.EndNanos,
-		)
-		if err != nil {
-			return nil, grpc.Errorf(codes.Internal, err.Error())
-		}
-		result := tspb.TimeSeriesQueryResponse_Result{
-			Query:      query,
-			Datapoints: datapoints,
+	workerOutput := make(chan queryResultWithError)
+
+	// Defer cancellation of context passed to worker tasks; if main task
+	// returns early, worker tasks should be torn down quickly.
+	ctx, cancel := context.WithCancel(ctx)
+	defer cancel()
+
+	// Start a task which is itself responsible for starting per-query worker
+	// tasks. This is needed because RunLimitedAsyncTask can block; in the
+	// case where a single request has more queries than the semaphore limit,
+	// a deadlock would occur because queries cannot complete until
+	// they have written their result to the "output" channel, which is
+	// processed later in the main function.
+	if err := s.stopper.RunAsyncTask(ctx, func(loopCtx context.Context) {
+		for queryNum := 0; queryNum < len(request.Queries); queryNum++ {
+			// If context has been canceled, there is no need to queue
+			// additional workers.
+			select {
+			case <-loopCtx.Done():
+				return
+			default:
+			}
+
+			queryNum := queryNum
+			if err := s.stopper.RunLimitedAsyncTask(
+				loopCtx,
+				s.workerSem,
+				func(workerCtx context.Context) {
+					datapoints, sources, err := s.db.Query(
+						workerCtx,
+						request.Queries[queryNum],
+						Resolution10s,
+						sampleNanos,
+						request.StartNanos,
+						request.EndNanos,
+					)
+
+					result := queryResultWithError{
+						queryNum: queryNum,
+					}
+					if err != nil {
+						result.err = grpc.Errorf(codes.Internal, err.Error())
+					} else {
+						result.result = tspb.TimeSeriesQueryResponse_Result{
+							Query:      request.Queries[queryNum],
+							Datapoints: datapoints,
+						}
+						result.result.Sources = sources
+					}
+
+					select {
+					case workerOutput <- result:
+					case <-workerCtx.Done():
+					}
+				},
+			); err != nil {
+				// Stopper has been closed and is draining. Return an error and
+				// exit the worker-spawning loop.
+				select {
+				case workerOutput <- queryResultWithError{
+					queryNum: queryNum,
+					err:      err,
+				}:
+				case <-loopCtx.Done():
+				}
+				return
+			}
 		}
+	}); err != nil {
+		return nil, err
+	}
 
-		result.Sources = sources
-		response.Results = append(response.Results, result)
+	response := tspb.TimeSeriesQueryResponse{
+		Results: make([]tspb.TimeSeriesQueryResponse_Result, len(request.Queries)),
+	}
+
+	completed := 0
+	for completed < len(request.Queries) {
```

Comment:
`for range request.Queries`. If the `ctx.Done()` case occurs you're returning and don't care about the index.